### PR TITLE
Update aggregator API for BYOHelper

### LINKED ISSUE #1506: Revisit aggregator API's `PostTaskReq` for self-service task provisioning
In support of the plot laid out in #1486, the aggregator API's task creation endpoint needs at least the following changes:

- Remove task ID (will be derived from VDAF verify key)
- VDAF verify key is required
- remove collector auth token (will be generated by leader Janus)

### PR DESCRIPTION
**Pull request by tgeoghegan**
Resolves #1506

### CODE REVIEW

**Comment by tgeoghegan** [comment]
This is working as far as the unit tests on `janus_aggregator_api` go, but next we need to test it against a `divviup-api` that's been updated to use the new `PostTaskReq`.

**Comment by tgeoghegan** [comment]
The test support code in `integration_tests/tests/in_cluster.rs` uses `divviup-api` for task provisioning. At the moment, it hits the `divviup-api` instance associated with either aggregator to separately provision tasks. However, `divviup-api` will soon assume responsibility for provisioning the task in both aggregators (https://github.com/divviup/divviup-api/issues/199). Once that lands (which will come with changes to the messages for task creation), we'll need a further Janus change to adopt those new messages. #1528

File: aggregator_api/src/lib.rs
Comment (tgeoghegan) [inline aggregator_api/src/lib.rs:431]: Ideally this structure would not contain a secret value like VDAF verify key, to reduce the risk of it being exposed. However I'm not sure if that breaks task sync.

**Comment by tgeoghegan** [comment]
Another thing we discussed is moving the aggregator API message definitions (`PostTaskReq`, `TaskResp`) into `janus_messages` so that `divviup-api` and others can avoid re-declaring them. That's a good idea, but it's going to require moving a whole bunch of type declarations between `janus_aggregator_core`, `janus_core` and `janus_messages`, so I'm not going to do that in this PR, both to avoid making this change any bigger, but also to avoid obscuring the changes to `PostTaskReq` and `TaskResp` that this change makes.

I think this change is ready for review so we can unblock further `divviup-api` work. We'll have to make sure to bump the versions used in `janus-ops` in lockstep once both sides are ready.

File: aggregator_api/src/lib.rs
Comment (tgeoghegan) [inline aggregator_api/src/lib.rs:185]: From discussion in Slack: we should incorporate a salt into the task ID derivation, since task ID is public(ish). One option is to have the client of the aggregator API (i.e. divviup-api) pass in the same salt to both aggregators when it provisions the task.

Comment (tgeoghegan) [reply aggregator_api/src/lib.rs:185]: After discussing further: the VDAF verify keys we currently deal with for Prio3 and Poplar1 are 16 bytes. The search space of 2^128 possible values is big enough that reversing the SHA-256 hash to get the VDAF verify key from a task ID is already impractical, so adding a salt isn't worth the complexity.

File: aggregator_api/src/lib.rs
Comment (inahga) [inline aggregator_api/src/lib.rs:187]: This strikes me as possibly InternalServerError? We've already validated that the vdaf_verify_key is of the correct length. Thus if we can't `try_from` then likely something went wrong with the `digest`, i.e. a bug.

File: aggregator_api/src/lib.rs
Comment (inahga) [inline aggregator_api/src/lib.rs:189]: Rust newbie question: what's the preference for `Vec::from([foo])` instead of `vec![foo]`? I would have leaned towards `vec!` but curious if there's something more subtle here.

File: aggregator_api/src/lib.rs
Comment (inahga) [inline aggregator_api/src/lib.rs:237]: Nit: In the spirit of DRY we could have a role check at the top of the function
```rust
if !matches(role, Role::Leader | Role::Helper) {
    /* same error logic */
}
```
Then subsequent matches over role can have the last case be `_ => unreachable!()`.

File: aggregator_api/src/lib.rs
Comment (inahga) [reply aggregator_api/src/lib.rs:433]: Refresh my memory on what task sync accomplishes? I thought it was to sync up task definitions from janus to control plane. If that's the case then presumably the leader and helper already have the vdaf_verify_key and handing this over is not necessary?

File: aggregator_api/src/lib.rs
Comment (inahga) [reply aggregator_api/src/lib.rs:461]: Where is it intended for the subscriber to get the collector auth token to run a collector?

**Review by inahga — COMMENTED** [review]
Mostly questions to check my understanding, a few improvements available.

File: aggregator_api/src/lib.rs
Comment (tgeoghegan) [reply aggregator_api/src/lib.rs:189]: I'm not sure if there's a Rust-wide convention for this, but we've used `Vec::from([T])` in Janus. There's nothing explicit about this in our [code style](https://github.com/divviup/janus/blob/main/docs/DEVELOPMENT.md#code-style), but I remember discussing this in the past -- @branlwyd do you remember if we wrote down anything about this style preference somewhere?

File: aggregator_api/src/lib.rs
Comment (tgeoghegan) [reply aggregator_api/src/lib.rs:425]: I think I agree. But I'm not certain whether `divviup-api` ever needs to see this value beyond provisioning it into the aggregators, so I'm just making a note that we have to think about this before we risk breaking task sync.

File: aggregator_api/src/lib.rs
Comment (tgeoghegan) [reply aggregator_api/src/lib.rs:453]: See [the sequence diagram](https://private-user-images.githubusercontent.com/19212995/244223772-4e0db56a-f55c-48de-9b09-dd15c0bed3b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJrZXkxIiwiZXhwIjoxNjg3MzcxNjc5LCJuYmYiOjE2ODczNzEzNzksInBhdGgiOiIvMTkyMTI5OTUvMjQ0MjIzNzcyLTRlMGRiNTZhLWY1NWMtNDhkZS05YjA5LWRkMTVjMGJlZDNiNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMDYyMVQxODE2MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYjQwNDUyMTVkMWUzY2Q3NmIyMzRiZjU4NjJhY2I2ZTYyOTdlNWI4NDE0MzZkYWQwOTI4MDgwYzRhYjA1ZWIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ngM2dTLOSF0ZuUDdLI9r6GWn34Qp2qTRWUYy_9OXC8c) in #1486. Short version:

- When the Leader is provisioning the task into itself, it generates the collector auth token and stores it in its own datastore.
- Leader delivers the collector auth token to divviup-api in this `TaskResp` message
- divviup-api then provides the collector auth token to the subscriber in the response to the `NewTask` message
- Collector is expected to store the auth token and use it later to authenticate collection job requests to the leader

The nuance here is that we need to give the subscriber the collector auth token "in the clear" when the task is created. But when we do #1509, Janus will start storing a token hash instead of the token. That means that a `GET /tasks/:id` request cannot contain the actual auth token, because the leader doesn't have it anymore at that point. I guess at that point this field can just be `None`, so maybe no message field changes will be needed.

File: aggregator_api/src/lib.rs
Comment (tgeoghegan) [reply aggregator_api/src/lib.rs]: Thanks, that is tidier

**Review by inahga — APPROVED** [review]

**Pull request merged**
